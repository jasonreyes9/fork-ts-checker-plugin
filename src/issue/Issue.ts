import {
  IssueSeverity,
  compareIssueSeverities,
  isIssueSeverity
} from './IssueSeverity';
import { IssueOrigin, compareIssueOrigins, isIssueOrigin } from './IssueOrigin';

interface Issue {
  origin: IssueOrigin;
  severity: IssueSeverity;
  code: string;
  message: string;
  file?: string;
  line?: number;
  character?: number;
  stack?: string;
}

function isIssue(value: unknown): value is Issue {
  return (
    !!value &&
    typeof value === 'object' &&
    isIssueOrigin((value as Issue).origin) &&
    isIssueSeverity((value as Issue).severity) &&
    !!(value as Issue).code &&
    !!(value as Issue).message
  );
}

function compareOptionalStrings(stringA?: string, stringB?: string) {
  if (stringA === stringB) {
    return 0;
  }

  if (stringA === undefined || stringA === null) {
    return -1;
  }
  if (stringB === undefined || stringB === null) {
    return 1;
  }

  return stringA.toString().localeCompare(stringB.toString());
}

function compareNumbers(numberA?: number, numberB?: number) {
  if (numberA === numberB) {
    return 0;
  }

  if (numberA === undefined || numberA === null) {
    return -1;
  }
  if (numberB === undefined || numberB === null) {
    return 1;
  }

  return Math.sign(numberA - numberB);
}

function compareIssues(issueA: Issue, issueB: Issue) {
  return (
    compareIssueOrigins(issueA.origin, issueB.origin) ||
    compareOptionalStrings(issueA.file, issueB.file) ||
    compareIssueSeverities(issueA.severity, issueB.severity) ||
    compareNumbers(issueA.line, issueB.line) ||
    compareNumbers(issueA.character, issueB.character) ||
    compareOptionalStrings(issueA.code, issueB.code) ||
    compareOptionalStrings(issueA.message, issueB.message) ||
    compareOptionalStrings(issueA.stack, issueB.stack) ||
    0 /* EqualTo */
  );
}

function equalsIssues(issueA: Issue, issueB: Issue) {
  return compareIssues(issueA, issueB) === 0;
}

function deduplicateAndSortIssues(issues: Issue[]) {
  const sortedIssues = issues.filter(isIssue).sort(compareIssues);

  return sortedIssues.filter(
    (issue, index) =>
      index === 0 || !equalsIssues(issue, sortedIssues[index - 1])
  );
}

export { Issue, isIssue, deduplicateAndSortIssues };
