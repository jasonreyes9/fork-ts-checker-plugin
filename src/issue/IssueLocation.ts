import { compareIssuePositions, IssuePosition } from './IssuePosition';

interface IssueLocation {
  start: IssuePosition;
  end: IssuePosition;
}

function compareIssueLocations(locationA?: IssueLocation, locationB?: IssueLocation) {
  if (locationA === locationB) {
    return 0;
  }

  if (!locationA) {
    return -1;
  }

  if (!locationB) {
    return 1;
  }

  return (
    compareIssuePositions(locationA.start, locationB.start) ||
    compareIssuePositions(locationA.end, locationB.end)
  );
}

function formatIssueLocation(location: IssueLocation) {
  return [
    `${location.start.line}:${location.start.column}`,
    location.start.line !== location.end.line
      ? `${location.end.line}:${location.end.column}`
      : `${location.end.column}`,
  ].join('-');
}

export { IssueLocation, compareIssueLocations, formatIssueLocation };
