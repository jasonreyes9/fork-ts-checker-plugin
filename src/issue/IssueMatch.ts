import { Issue } from './index';
import { IssuePredicate } from './IssuePredicate';
import minimatch from 'minimatch';
import path from 'path';
import normalizeSlash from '../utils/path/normalizeSlash';

type IssueMatch = Partial<Pick<Issue, 'origin' | 'severity' | 'code' | 'file'>>;

function createIssuePredicateFromIssueMatch(context: string, match: IssueMatch): IssuePredicate {
  return (issue) => {
    const matchesOrigin = !match.origin || match.origin === issue.origin;
    const matchesSeverity = !match.severity || match.severity === issue.severity;
    const matchesCode = !match.code || match.code === issue.code;
    const matchesFile =
      !issue.file ||
      (!!issue.file &&
        (!match.file || minimatch(normalizeSlash(path.relative(context, issue.file)), match.file)));

    return matchesOrigin && matchesSeverity && matchesCode && matchesFile;
  };
}

export { IssueMatch, createIssuePredicateFromIssueMatch };
