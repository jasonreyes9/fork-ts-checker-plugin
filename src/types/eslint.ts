// copy eslint types as installation of eslint package is optional
export interface LintMessage {
  column: number;
  line: number;
  endColumn?: number;
  endLine?: number;
  ruleId: string | null;
  message: string;
  nodeType: string;
  fatal?: true;
  severity: 0 | 1 | 2;
  fix?: {
    range: [number, number];
    text: string;
  };
  source: string | null;
}

export interface LintResult {
  filePath: string;
  messages: LintMessage[];
  errorCount: number;
  warningCount: number;
  fixableErrorCount: number;
  fixableWarningCount: number;
  output?: string;
  source?: string;
}

export interface LintReport {
  results: LintResult[];
  errorCount: number;
  warningCount: number;
  fixableErrorCount: number;
  fixableWarningCount: number;
}

export interface Options {
  fix?: boolean;

  // The rest of the properties are not specified here because they are not
  // directly used by this package and are instead just passed to eslint.
  // We do this in order to avoid a dependency on @types/eslint (since the use
  // of eslint is optional) and to avoid copying types from @types/eslint.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}
