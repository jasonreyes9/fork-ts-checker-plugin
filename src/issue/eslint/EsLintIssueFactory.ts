import * as eslint from 'eslint';
import { FileAwareEsLintMessage } from './FileAwareEsLintMessage';
import { deduplicateAndSortIssues, Issue } from '../Issue';
import { IssueOrigin } from '../IssueOrigin';
import { IssueSeverity } from '../IssueSeverity';

function createIssueFromEsLintMessage(message: FileAwareEsLintMessage): Issue {
  return {
    origin: IssueOrigin.ESLINT,
    code: message.ruleId ? String(message.ruleId) : '[unknown]',
    severity:
      message.severity === 1 ? IssueSeverity.WARNING : IssueSeverity.ERROR,
    message: message.message,
    file: message.filePath,
    line: message.line,
    character: message.column
  };
}

function createFileAwareEsLintMessagesFromEsLintResult(
  result: eslint.CLIEngine.LintResult
): FileAwareEsLintMessage[] {
  return result.messages.map(message => ({
    ...message,
    filePath: result.filePath
  }));
}

function createFileAwareEsLintMessagesFromEsLintReport(
  report: eslint.CLIEngine.LintReport
): FileAwareEsLintMessage[] {
  return report.results.reduce<FileAwareEsLintMessage[]>(
    (messages, result) => [
      ...messages,
      ...createFileAwareEsLintMessagesFromEsLintResult(result)
    ],
    []
  );
}

function createFileAwareEsLintMessagesFromEsLintReports(
  reports: eslint.CLIEngine.LintReport[]
): FileAwareEsLintMessage[] {
  return reports.reduce<FileAwareEsLintMessage[]>(
    (messages, report) => [
      ...messages,
      ...createFileAwareEsLintMessagesFromEsLintReport(report)
    ],
    []
  );
}

function createIssuesFromEsLintMessages(
  messages: FileAwareEsLintMessage[]
): Issue[] {
  return deduplicateAndSortIssues(messages.map(createIssueFromEsLintMessage));
}

function createIssuesFromEsLintReports(
  reports: eslint.CLIEngine.LintReport[]
): Issue[] {
  return createIssuesFromEsLintMessages(
    createFileAwareEsLintMessagesFromEsLintReports(reports)
  );
}

export { createIssuesFromEsLintReports };
