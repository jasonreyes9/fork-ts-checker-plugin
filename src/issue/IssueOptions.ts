import { IssueMatch } from './IssueMatch';
import { IssuePredicate } from './IssuePredicate';

type IssuePredicateOption = IssuePredicate | IssueMatch | (IssuePredicate | IssueMatch)[];

interface IssueOptions {
  include?: IssuePredicateOption;
  exclude?: IssuePredicateOption;
  scope?: 'all' | 'webpack';
}

export { IssueOptions, IssuePredicateOption };
