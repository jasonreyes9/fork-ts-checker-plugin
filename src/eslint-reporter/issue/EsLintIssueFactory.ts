import { Issue, IssueLocation } from '../../issue';
import { LintMessage, LintResult } from '../types/eslint';

function createIssueFromEsLintMessage(filePath: string, message: LintMessage): Issue {
  let location: IssueLocation | undefined;

  if (message.line) {
    location = {
      start: {
        line: message.line,
        column: message.column,
      },
      end: {
        line: message.endLine || message.line,
        column: message.endColumn || message.column,
      },
    };
  }

  return {
    origin: 'eslint',
    code: message.ruleId ? String(message.ruleId) : '[unknown]',
    severity: message.severity === 1 ? 'warning' : 'error',
    message: message.message,
    file: filePath,
    location,
  };
}

function createIssuesFromEsLintResults(results: LintResult[]): Issue[] {
  return results.reduce<Issue[]>(
    (messages, result) => [
      ...messages,
      ...result.messages.map((message) => createIssueFromEsLintMessage(result.filePath, message)),
    ],
    []
  );
}

export { createIssuesFromEsLintResults };
