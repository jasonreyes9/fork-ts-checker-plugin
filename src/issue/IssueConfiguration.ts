import webpack from 'webpack';
import { createIssuePredicateFromIssueMatch } from './IssueMatch';
import {
  composeIssuePredicates,
  createTrivialIssuePredicate,
  IssuePredicate,
} from './IssuePredicate';
import { IssuePredicateOption, IssueOptions } from './IssueOptions';

interface IssueConfiguration {
  predicate: IssuePredicate;
}

function createIssuePredicateFromOption(
  context: string,
  option: IssuePredicateOption
): IssuePredicate {
  if (Array.isArray(option)) {
    return composeIssuePredicates(
      option.map((option) =>
        typeof option === 'function' ? option : createIssuePredicateFromIssueMatch(context, option)
      )
    );
  }

  return typeof option === 'function'
    ? option
    : createIssuePredicateFromIssueMatch(context, option);
}

function createIssueConfiguration(
  compiler: webpack.Compiler,
  options: IssueOptions | undefined
): IssueConfiguration {
  const context = compiler.options.context || process.cwd();

  if (!options) {
    options = {} as IssueOptions;
  }

  const include = options.include
    ? createIssuePredicateFromOption(context, options.include)
    : createTrivialIssuePredicate(true);
  const exclude = options.exclude
    ? createIssuePredicateFromOption(context, options.exclude)
    : createTrivialIssuePredicate(false);

  return {
    predicate: (issue) => include(issue) && !exclude(issue),
  };
}

export { IssueConfiguration, createIssueConfiguration };
