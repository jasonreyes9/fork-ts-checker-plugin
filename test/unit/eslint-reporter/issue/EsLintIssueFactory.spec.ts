import { LintMessage, LintResult } from '../../../../lib/types/eslint';
import { createIssuesFromEsLintResults } from '../../../../lib/eslint-reporter/issue/EsLintIssueFactory';

describe('eslint-reporter/issue/EsLintIssueFactory', () => {
  const ES_LINT_MESSAGE_ERROR: LintMessage = {
    column: 0,
    line: 13,
    endColumn: 5,
    endLine: 13,
    ruleId: 'no-unused-vars',
    message: `'y' is assigned a value but never used.`,
    nodeType: '',
    severity: 0,
    source: null,
  };
  const ES_LINT_MESSAGE_WARNING: LintMessage = {
    column: 10,
    line: 15,
    ruleId: 'no-unused-vars',
    message: `'y' is assigned a value but never used.`,
    nodeType: '',
    severity: 1,
    source: null,
  };
  const ES_LINT_RESULT_INDEX: LintResult = {
    filePath: 'src/index.ts',
    messages: [ES_LINT_MESSAGE_ERROR],
    errorCount: 1,
    warningCount: 0,
    fixableErrorCount: 0,
    fixableWarningCount: 0,
  };
  const ES_LINT_RESULT_ANOTHER: LintResult = {
    filePath: 'src/another.ts',
    messages: [ES_LINT_MESSAGE_WARNING],
    errorCount: 0,
    warningCount: 1,
    fixableErrorCount: 0,
    fixableWarningCount: 0,
  };
  const ES_LINT_RESULT_ADDITIONAL: LintResult = {
    filePath: 'src/additional.ts',
    messages: [ES_LINT_MESSAGE_ERROR],
    errorCount: 1,
    warningCount: 0,
    fixableErrorCount: 0,
    fixableWarningCount: 0,
  };
  const ES_LINT_RESULTS = [ES_LINT_RESULT_INDEX, ES_LINT_RESULT_ANOTHER, ES_LINT_RESULT_ADDITIONAL];

  it.each([[ES_LINT_RESULTS]])('creates Issues from EsLint Results: %p', (results) => {
    const issues = createIssuesFromEsLintResults(results);

    expect(issues).toMatchSnapshot();
  });
});
