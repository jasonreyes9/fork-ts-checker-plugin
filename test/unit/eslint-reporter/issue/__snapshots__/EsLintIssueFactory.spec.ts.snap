// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eslint-reporter/issue/EsLintIssueFactory creates Issues from EsLint Results: [[Object], [Object], [Object]] 1`] = `
Array [
  Object {
    "code": "no-unused-vars",
    "file": "src/index.ts",
    "location": Object {
      "end": Object {
        "column": 5,
        "line": 13,
      },
      "start": Object {
        "column": 0,
        "line": 13,
      },
    },
    "message": "'y' is assigned a value but never used.",
    "origin": "eslint",
    "severity": "error",
  },
  Object {
    "code": "no-unused-vars",
    "file": "src/another.ts",
    "location": Object {
      "end": Object {
        "column": 10,
        "line": 15,
      },
      "start": Object {
        "column": 10,
        "line": 15,
      },
    },
    "message": "'y' is assigned a value but never used.",
    "origin": "eslint",
    "severity": "warning",
  },
  Object {
    "code": "no-unused-vars",
    "file": "src/additional.ts",
    "location": Object {
      "end": Object {
        "column": 5,
        "line": 13,
      },
      "start": Object {
        "column": 0,
        "line": 13,
      },
    },
    "message": "'y' is assigned a value but never used.",
    "origin": "eslint",
    "severity": "error",
  },
]
`;
